internal/mux/endpoint.go

```diff
@@ -0,0 +1,67 @@
+package mux
+
+import (
+	"errors"
+	"net"
+	"time"
+)
+
+// Endpoint implements net.Conn. It is used to read muxed packets.
+type Endpoint struct {
+	mux    *Mux
+	readCh chan []byte
+	sizeCh chan int
+	doneCh chan struct{}
+}
+
+// Close unregisters the endpoint from the Mux
+func (e *Endpoint) Close() error {
+	e.mux.RemoveEndpoint(e)
+	return nil
+}
+
+func (e *Endpoint) close() {
+	close(e.doneCh)
+}
+
+// Read reads a packet of len(p) bytes from the underlying conn
+// that are matched by the associated MuxFunc
+func (e *Endpoint) Read(p []byte) (int, error) {
+	select {
+	case e.readCh <- p:
```

Comment:
I wish channels could have a different datatype for in/out (so we don't need two unique channels)

This might be a nit, but do you think we could choose a better name. When I first saw this pattern I assumed `readCh` and `sizeCh` were tied to each other (like `readCh` returns data and `sizeCh` returns how much) since we are going to use this a lot might be worth making it less confusing.

Reply:
Any suggestions? I'm not sure how to convey the entire idea in the channel names.